draft-ietf-tls-esni.md

```diff
@@ -739,10 +739,15 @@ for servers to proceed with the connection and rely on the client to abort if
 ESNI was required. In particular, the unrecognized value alone does not
 indicate a misconfigured ESNI advertisement ({{misconfiguration}}). Instead,
 servers can measure occurrences of the "esni_required" alert to detect this
-case.
+case. An empty ClientEncryptedSNI.record_digest value MAY be used in environments
+wherein trial decryption is a viable approach for matching ClientEncryptedSNI
+contents to a known ESNIKeys. (Some uses of ESNI, such as local discovery mode,
+may opt to omit the ClientEncryptedSNI.record_digest since it can be used as a
+tracking vector.)
 
-If the ClientEncryptedSNI.record_digest value does match the cryptographic
-hash of a known ESNIKeys, the server performs the following checks:
+If the ClientEncryptedSNI.record_digest value does match a known ESNIKeys, either by
+comparing against a known cryptographic hash or by trial decryption, the server
+performs the following checks:
```

Comment:
To clarify, is the intent here is that servers outside those environments are not required to implement trial decryption?

The text reads a little unclear to me. The paragraph above just says an empty one "MAY be used in [some environments]", but that could mean by either the client or the server. This paragraph also could read as if the server needs to check the trial decryption, otherwise it doesn't know whether to follow these steps.

(Trial decryption would more fully solve the issue #154 is trying to address, but I worry about the performance cost of asymmetric trial decryption if you rotate keys decently often. Then again, the robustness fixes do mean it's okay to drop older keys sooner if only used by a tiny fraction of clients...)

Reply:
The trial decryption should be opt-in. The typical use case is in local discovery, with DTLS or QUIC. Broadcast or multicast the first packet, and then all servers who are ready to be discovered do trial detection. Yes, this is trading off CPU for server privacy, which is only OK in some scenario. In a typical implementation, the server would have a "discovery" flag, which will enable trial decryption when "waiting to be discovered".

Reply:
> To clarify, is the intent here is that servers outside those environments are not required to implement trial decryption?

Precisely. I see what you mean about it implying trial decryption is necessary. Perhaps: 

```
If the ClientEncryptedSNI.record_digest value does match a known ESNIKeys, the server...
```

And then later on...

```
If ClientEncryptedSNI.record_digest is non-empty, servers SHOULD compare it against cryptographic hashes of known ESNIKeys to find a match. If ClientEncryptedSNI.record_digest is empty, servers MAY use trial decryption to match to a known ESNIKeys.
```

Reply:
@davidben @huitema: I made an attempt at clarifying the text. Suggestions for improvements are welcome!